File: ts/project/.sdk/tm/c/utility/param.c
// param utility (mirrors utility/param.rs).

#include "sdk.h"

#include <string.h>

voxgig_value* param_util(Context* ctx, voxgig_value* paramdef) {
  voxgig_value* point = ctx->point;
  Spec* spec = ctx->spec;
  voxgig_value* mtch = ctx->mtch;
  voxgig_value* reqmatch = ctx->reqmatch;
  voxgig_value* data = ctx->data;
  voxgig_value* reqdata = ctx->reqdata;

  int pt = voxgig_typify(paramdef);

  char key[256];
  key[0] = '\0';
  if (0 != ((VOXGIG_T_STRING) & pt)) {
    if (voxgig_is_string(paramdef)) {
      const char* s = voxgig_as_string(paramdef);
      snprintf(key, sizeof(key), "%s", s);
    }
  } else {
    const char* n = get_str(paramdef, "name");
    if (n) snprintf(key, sizeof(key), "%s", n);
  }

  char akey[256];
  akey[0] = '\0';
  if (!v_is_noval(point)) {
    voxgig_value* alias = to_map(getp(point, "alias"));
    if (!v_is_noval(alias)) {
      const char* ak = get_str(alias, key);
      if (ak) snprintf(akey, sizeof(akey), "%s", ak);
    }
  }

  voxgig_value* val = getp(reqmatch, key);
  if (v_is_noval(val)) val = getp(mtch, key);

  if (v_is_noval(val) && akey[0] != '\0') {
    if (spec) {
      setp(spec->alias, akey, v_str(key));
    }
    val = getp(reqmatch, akey);
  }

  if (v_is_noval(val)) val = getp(reqdata, key);
  if (v_is_noval(val)) val = getp(data, key);

  if (v_is_noval(val) && akey[0] != '\0') {
    val = getp(reqdata, akey);
    if (v_is_noval(val)) val = getp(data, akey);
  }

  return val;
}

File: ts/project/.sdk/tm/rust/utility/param.rs
use std::rc::Rc;

use crate::core::context::Context;
use crate::core::helpers::{get_str, getp, to_map};
use crate::utility::voxgigstruct as vs;
use crate::utility::voxgigstruct::Value;

pub fn param_util(ctx: &Rc<Context>, paramdef: &Value) -> Value {
    let point = ctx.point.borrow().clone();
    let spec = ctx.spec.borrow().clone();
    let mtch = ctx.mtch.borrow().clone();
    let reqmatch = ctx.reqmatch.borrow().clone();
    let data = ctx.data.borrow().clone();
    let reqdata = ctx.reqdata.borrow().clone();

    let pt = vs::typify(paramdef);

    let key = if 0 != ((vs::T_STRING as i64) & pt) {
        match paramdef {
            Value::Str(s) => s.clone(),
            _ => String::new(),
        }
    } else {
        get_str(paramdef, "name").unwrap_or_default()
    };

    let mut akey = String::new();
    if !point.is_noval() {
        let alias = to_map(&getp(&point, "alias"));
        if !alias.is_noval() {
            if let Some(ak) = get_str(&alias, &key) {
                akey = ak;
            }
        }
    }

    let mut val = getp(&reqmatch, &key);

    if val.is_noval() {
        val = getp(&mtch, &key);
    }

    if val.is_noval() && !akey.is_empty() {
        if let Some(sp) = &spec {
            let alias = sp.borrow().alias.clone();
            crate::core::helpers::setp(&alias, &akey, Value::str(key.clone()));
        }
        val = getp(&reqmatch, &akey);
    }

    if val.is_noval() {
        val = getp(&reqdata, &key);
    }

    if val.is_noval() {
        val = getp(&data, &key);
    }

    if val.is_noval() && !akey.is_empty() {
        val = getp(&reqdata, &akey);
        if val.is_noval() {
            val = getp(&data, &akey);
        }
    }

    val
}
